### PR TITLE
fix: reclassify Image, Graph, Reasoning, Consensus engines to advisory-only (#257)

### LINKED ISSUE #134: [Bug]: ImageVerifier allows VLM fallback to decide final verification verdicts
### QWED Version

5.0.0

### Python Version

3.11.9

### Operating System

Windows

### Which engine is affected?

Image Verifier

### Input that caused the bug

```python
from qwed_new.core.image_verifier import ImageVerifier

verifier = ImageVerifier(use_vlm_fallback=True)

# Any image/claim pair that forces the engine into VLM_REQUIRED
image_bytes = b"fake-image-bytes-for-example"
claim = "The person in the image is wearing a police uniform."

result = verifier.verify_image(
    image_bytes=image_bytes,
    claim=claim,
)

print(result)


### Expected behavior

When deterministic image checks cannot prove a claim and the engine enters a `VLM_REQUIRED` state, the verifier must fail closed.

**Expected behavior:**
- `VLM_REQUIRED` must not be upgraded into a final verification verdict by a VLM
- external model output may be recorded as advisory analysis only
- if the claim cannot be deterministically proven, the final state must remain non-pass such as `INCONCLUSIVE` or `UNVERIFIED`

**QWED principle:**
A model may interpret an image, but it must not become the proof boundary.


### Actual behavior

`ImageVerifier.verify_image()` currently allows VLM fallback output to replace the engine’s result when the deterministic path returns `VLM_REQUIRED`.

Relevant code in `src/qwed_new/core/image_verifier.py`:

```python
if result.verdict == "VLM_REQUIRED" and self.use_vlm_fallback and self.vlm_provider:
    methods_used.append("vlm_analysis")
    vlm_result = self._vlm_fallback(image_bytes, claim)
    if vlm_result:
        result = ImageVerificationResult(
            verdict=vlm_result.get("verdict", "INCONCLUSIVE"),
            confidence=vlm_result.get("confidence", 0.5) * 0.8,
            reasoning=vlm_result.get("reasoning", "VLM analysis")
        )
```

**This means:**
- the deterministic verifier explicitly says visual proof is not available locally
- a VLM is then allowed to supply the final verdict
- the engine’s final verification result can become model-decided

## Severity
HIGH

## Why this violates QWED philosophy
QWED requires:
- fail-closed verification
- zero trust in model outputs
- no heuristic acceptance of results
- explicit proof, not inferred truth

`VLM_REQUIRED` means the engine does not have deterministic proof capability for that claim. Allowing a VLM to replace that state with a concrete verdict converts “needs interpretation” into “verified by model opinion.”

That is a trust-boundary breach.

## Attack / bypass scenario
An attacker can submit claims that are hard to deterministically verify but easy to bias visually, such as:
- identity or role claims
- object presence claims
- intent or scene interpretation claims
- context-heavy visual semantics

Once the engine enters `VLM_REQUIRED`, the fallback model can emit `SUPPORTED` / `REFUTED` / similar verdicts that are surfaced as if they were verification outcomes.

That allows ambiguous visual interpretation to cross into the final verification result.

## Fix direction
Do not allow VLM fallback to replace the final verifier verdict.

Required direction:
1. preserve `VLM_REQUIRED` / `INCONCLUSIVE` as the final non-pass verification state
2. store VLM output only as advisory analysis or side-channel metadata
3. do not expose VLM interpretation as a proof-bearing verdict
4. separate deterministic verification status from model-assisted interpretation status

## Suggested regression tests
- `VLM_REQUIRED` path + VLM available -> final verifier verdict must remain non-pass
- VLM advisory output may be attached, but must not replace deterministic engine status
- no-VLM path must still remain `INCONCLUSIVE`
- complex visual claims must never become `VERIFIED` solely because a VLM answered confidently

## QWED Principle To Preserve
If deterministic visual proof is unavailable, the verifier must stop at uncertainty instead of outsourcing truth to a model.

### LINKED ISSUE #257: [Feature]: Reclassify Image, Graph, Reasoning, and Consensus engines to advisory-only (Phase 3)
## Purpose

Reclassify ImageVerifier, GraphFactVerifier, ReasoningVerifier, and ConsensusVerifier to advisory-only mode — they must never return VERIFIED. These engines mix deterministic and heuristic signals. Advisory/heuristic signals must be carried as AdvisoryCheck in developer_fields, never promoted to VERIFIED status.

## Related

- **Meta:** #216 (Phase 3 — Hybrid engines)
- **Bug:** #134 — ImageVerifier allows VLM fallback to decide final verification verdicts
- **Bug:** #163 — GraphFactVerifier can return VERIFIED from probabilistic or partial support
- **Bug:** #164 — ReasoningVerifier can return valid results without an actual proof or provider path
- **Bug:** #135 — ConsensusVerifier fabricates fallback math values instead of failing closed on translation gaps

### 1. ImageVerifier (#134)

- VLM_REQUIRED + VLM available → VERIFIED is the bug
- Fix: VLM output goes to advisory_checks only, final status remains UNVERIFIABLE
- Deterministic paths (dimension match) can still produce VERIFIED with proof_ref
- confidence from VLM → developer_fields.advisory_checks.vlm_confidence (never verdict-deciding)
- VLM reasoning → developer_fields.advisory_checks.vlm_reasoning (advisory_only=True)

### 2. GraphFactVerifier (#163)

- 50% threshold → VERIFIED is the bug
- Fix: all material claim triples must be supported for VERIFIED
- Partial support → UNVERIFIABLE with coverage ratio in developer_fields
- NLI fallback → advisory_checks only, never verdict-deciding
- 0.5 threshold must surface as constraint_id in developer_fields

### 3. ReasoningVerifier (#164)

- is_valid = len(issues) == 0 without proof path is the bug
- Fix: no provider/proof path → UNVERIFIABLE
- Heuristic consistency checks → advisory_checks only
- "No LLM provider available" → UNVERIFIABLE, not valid

### 4. ConsensusVerifier (#135 + confidence collapse audit finding)

- Fabricated fallback math values (regex extraction, default to 0) → must fail closed
- Fix: _parse_math_query() must return UNVERIFIABLE/BLOCKED when translation fails
- Missing expected_value → BLOCKED, not defaulted to 0
- Fabricated tasks must never reach downstream engines

### 5. Confidence collapse (audit finding from #216)

consensus_verifier.py:564-767 collapses all engine statuses to a single confidence: 0.0-1.0 score. BLOCKED, UNVERIFIABLE, and ERROR all map to confidence: 0.0. The orchestrator cannot distinguish "couldn't attempt" (BLOCKED) from "tried but inconclusive" (UNVERIFIABLE) from "execution failed" (ERROR).

Fix: Add status-preserving consensus that maintains DiagnosticResult status through aggregation rather than collapsing to confidence.

## AdvisoryCheck Pattern (all 4 engines)

All heuristic/advisory signals must follow this pattern:

```python
developer_fields = {
    "advisory_checks": {
        "source": "vlm" | "nli" | "llm_fallback" | "heuristic",
        "verdict": "SUPPORTED" | "REFUTED" | ...,
        "confidence": 0.0-1.0,
        "reasoning": "...",
        "advisory_only": True,  # structural enforcement
        "constraint_id": "engine_name.advisory_only",
    },
}
```

## Success Criteria

- ImageVerifier: VLM path → UNVERIFIABLE (not VERIFIED)
- GraphFactVerifier: Partial support → UNVERIFIABLE (not VERIFIED)
- ReasoningVerifier: No provider → UNVERIFIABLE (not VERIFIED)
- ConsensusVerifier: Translation failure → BLOCKED (not fabricated)
- All advisory signals in developer_fields.advisory_checks with advisory_only=True
- Consensus preserves DiagnosticStatus through aggregation (no confidence collapse)
- proof_ref is null for all advisory-only paths
- Regression tests assert on DiagnosticResult fields for every fail path

### PR DESCRIPTION
**Pull request by rahuldass19**
## Summary

Make ImageVerifier, GraphFactVerifier, ReasoningVerifier, and ConsensusVerifier advisory-only � heuristic/model fallback paths never return VERIFIED (always UNVERIFIABLE or BLOCKED).

Closes #257. Fixes #134, #163, #164, #135.

### ImageVerifier (image_verifier.py)
- VLM path ? returns UNVERIFIABLE with advisory-only advisory checks
- Deterministic paths (size/type checks) ? VERIFIED with proof_ref from pixel evidence

### GraphFactVerifier (graph_fact_verifier.py)
- Threshold raised from 0.5 ? 1.0 (all material triples must match)
- verify_with_nli() ? returns UNVERIFIABLE with NLI results in advisory checks

### ReasoningVerifier (reasoning_verifier.py)
- No-provider path ? UNVERIFIABLE with constraint_id: reasoning_verifier.no_provider (fixes is_valid=len(issues)==0 false-positive)
- providers=[] no longer silently defaults to ["anthropic"]

### ConsensusVerifier (consensus_verifier.py)
- _parse_math_query() raises ValueError on translation failure (no regex/decimal fabrication)
- EngineResult.status field added
- _calculate_consensus() preserves BLOCKED ? BLOCKED, UNVERIFIABLE ? UNVERIFIABLE, VERIFIED ? VERIFIED

## Testing
- 20 new regression tests in tests/security/test_hybrid_advisory_only.py
- 32 existing graph_fact tests updated to DiagnosticResult field access pattern
- Full suite: 158 tests passing

## Summary by CodeRabbit

* **New Features**
  * Unified diagnostic result handling across image, graph-fact, reasoning, and consensus verification, including structured evidence, constraint IDs, and advisory checks.
  * Advisory-only model/heuristic paths no longer produce a top-level verified outcome on their own.
* **Bug Fixes**
  * Math translation errors are now fail-closed (blocked) instead of falling back.
  * Graph-fact verification is stricter (near-exact triple support required); word-boundary entity matching and NLI becomes advisory-only.
  * Consensus now propagates blocked/unverifiable diagnostics and improves async failure handling.
* **Tests**
  * Expanded security and regression coverage for advisory-only guarantees and edge-case status propagation.

___

## **CodeAnt-AI Description**
Make model- and heuristic-based verification advisory-only while preserving reliable deterministic proof

### What Changed
- Image checks now return `VERIFIED` only for deterministic evidence, `UNVERIFIABLE` for visual-model advice, and `BLOCKED` when deterministic data refutes a claim.
- Graph verification now requires every material fact to have near-exact support; partial matches and NLI results remain inconclusive or advisory.
- Reasoning verification no longer treats missing providers or heuristic agreement as proof, and explicitly reports unavailable-provider outcomes.
- Consensus verification no longer fabricates answers when translation fails, preserves blocked and inconclusive engine statuses, and requires unanimous unblocked agreement to verify.
- Verification responses now use a consistent diagnostic format with evidence, advisory checks, constraints, and developer details; related API and tests were updated accordingly.
- Entity and predicate matching avoids accidental substring matches such as “Tim” in “Timothy” or “is” in “visits.”

### Impact
`✅ Fewer false-positive verifications`
`✅ Clearer blocked and inconclusive outcomes`
`✅ Reliable deterministic proof boundaries`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

### CODE REVIEW

**Comment by rahuldass19** [comment]
@coderabbit review

File: src/qwed_new/core/graph_fact_verifier.py [hunk 1]
```diff
@@ -432,28 +446,20 @@ def _match_triples(
...
-        """Check if two entities match (exact or containment)."""
+        """Check if two entities match (exact or word-boundary containment)."""
         # Exact match
         if entity1 == entity2:
             return True
         
-        # Containment (one contains the other)
-        if entity1 in entity2 or entity2 in entity1:
+        # Word-boundary containment (e.g. "Modi" in "Narendra Modi" but NOT "Tim" in "Timothy")
+        if _word_boundary_contained(entity1, entity2) or _word_boundary_contained(entity2, entity1):
```
[the remark this answers is not in the document]
Comment (rahuldass19) [reply src/qwed_new/core/graph_fact_verifier.py:455]: This is a pre-existing design tradeoff, not a regression from this PR. The containment matching for whole-word substrings (e.g., "Tim" in "Tim Cook") is the same mechanism that makes "Modi" in "Narendra Modi" work — a core use case validated by test_containment_match. The Greptile fix was specifically for non-word-boundary substrings ("Tim" in "Timothy"), which bare in incorrectly allowed. Distinguishing "Tim" as a first name from "Modi" as a last-name fragment requires context-aware NLP, which is outside the deterministic scope of this engine. We can file a separate issue if real-world false positives appear.
